### PR TITLE
Refactor kubernetes Ansible role

### PR DESCRIPTION
**Pull request by maratoid**
1) It breaks up the kubernetes ansible role into several smaller roles with dependencies through meta/main.yaml
2) Gets rid of the many bash scripts ran as systemd services, and converts those into ansible tasks
3) removes variables from ansible inventory, putting them instead into group_vars files
4) removes sed-based post-processing of values in kraken-services yamls, in favor of using jinja2

This will most likely not pass the Jenkins muster, as it requires:
1) https://github.com/Samsung-AG/kraken-services/pull/77 to be merged first
2) new ansible-in-docker image to be pushed out to quay

### CODE REVIEW

File: ansible/ansible_in_docker.yaml [hunk 1]
```diff
@@ -1,9 +1,22 @@
...
   roles:
     - { role: ansible-coreos-bootstrap, coreos_pypy_url: 'https://bitbucket.org/squeaky/portable-pypy/downloads' }
 
-- hosts: etcd:master:nodes:apiservers
+- hosts: etcd
   roles:
-    - kubernetes
\ No newline at end of file
+    - { role: cadvisor }
+    - { role: kraken-common }
```
Comment (spiffxp) [inline ansible/ansible_in_docker.yaml]: maybe wrap this up in a `kraken-etcd` role?

File: ansible/ansible_in_docker.yaml [hunk 2]
```diff
@@ -1,9 +1,22 @@
 ---
-- hosts: etcd:master:nodes:apiservers
+- hosts: etcd:master:node:apiserver
```
Comment (spiffxp) [inline ansible/ansible_in_docker.yaml]: maybe call this playbook `kraken` instead of `ansible_in_docker`?

Comment (maratoid) [inline ansible/ansible_in_docker.yaml]: agree. this is now usable from both docker container and local dev machine.

File: ansible/ansible_in_docker.yaml (hunk 1 above)
Comment (maratoid) [inline ansible/ansible_in_docker.yaml]: makes sense

File: ansible/roles/kraken-master/tasks/thirdparty-scheduler.yaml [hunk 3]
```diff
@@ -16,8 +16,8 @@
     /opt/bin/kubectl --kubeconfig={{ kubelet.kubeconfig | expanduser }} create -f /opt/bin/kraken-services-rendered/{{ thirdparty_scheduler }} || /opt/bin/kubectl --kubeconfig={{ kubelet.kubeconfig | expanduser }} apply -f /opt/bin/kraken-services-rendered/{{ thirdparty_scheduler }}
   register: pod_result
   until: pod_result.rc == 0
-  retries: 100
-  delay: 5
+  retries: 120
+  delay: 10
 
 - name: Shut down binary scheduler
```
Comment (spiffxp) [inline ansible/roles/kraken-master/tasks/thirdparty-scheduler.yaml]: Missed a few `docker` to `container` renames in this file

File: ansible/roles/k8s-auth/tasks/create_auth.yaml [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+- name: Fill out credentails file for system users
+  become: yes
+  lineinfile: dest={{ kubernetes_cert_dir }}/known_tokens.csv
+                regexp=''
+                insertafter=EOF
+                line="{{ item.value.token }},{{ item.value.name }},{{ item.value.name }}"
+  with_dict: "{{ system_users }}"
+
+- name: Remove credentails csv generated for basic auth user
```
Comment (spiffxp) [inline ansible/roles/k8s-auth/tasks/create_auth.yaml]: Why are we removing this file and then `touch`/`lineinfile`'ing it?

File: ansible/roles/k8s-auth/tasks/create_auth.yaml [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  become: yes
+  file: path={{ kubernetes_cert_dir }}/basic_auth.csv state=absent
+
+- name: Remove credentails htpasswd generated for basic auth user
```
Comment (spiffxp) [inline ansible/roles/k8s-auth/tasks/create_auth.yaml]: Ditto for this file... guessing this is just a straight port of what the previous shell script was doing.  Just don't follow why we'd want to rewrite these files every playbook run

File: ansible/roles/k8s-auth/tasks/create_auth.yaml (hunk 4 above)
Comment (maratoid) [inline ansible/roles/k8s-auth/tasks/create_auth.yaml]: as far as i can tell lininfile only appends

File: ansible/roles/k8s-auth/tasks/create_auth.yaml (hunk 5 above)
Comment (maratoid) [inline ansible/roles/k8s-auth/tasks/create_auth.yaml]: again, as far as i can tell lininfile only appends. also there is a guard preventing regeneration of auth on master at main yaml level.

File: ansible/roles/kraken-master/tasks/thirdparty-scheduler.yaml (hunk 3 above)
Comment (maratoid) [inline ansible/roles/kraken-master/tasks/thirdparty-scheduler.yaml]: indeed

Comment (maratoid) [inline ansible/roles/kraken-master/tasks/thirdparty-scheduler.yaml]: also in variable descriptions, in terraform

File: ansible/roles/k8s-auth/tasks/create_auth.yaml (hunk 4 above)
Comment (spiffxp) [inline ansible/roles/k8s-auth/tasks/create_auth.yaml]: You can iterate over dicts in templates FWIW

File: terraform/aws/aws.tf [hunk 6]
```diff
@@ -485,10 +488,10 @@ resource "template_file" "node_cloudinit_special" {
     kubernetes_cert_dir        = "${var.kubernetes_cert_dir}"
     logentries_token           = "${var.logentries_token}"
     logentries_url             = "${var.logentries_url}"
-    master_port                = "${var.master_port}"
+    access_port                = "${var.access_port}"
     master_private_ip          = "${aws_instance.kubernetes_master.private_ip}"
     master_public_ip           = "${aws_instance.kubernetes_master.public_ip}"
-    master_scheme              = "${var.master_scheme}"
+    access_scheme              = "${var.access_scheme}"
```
Comment (spiffxp) [inline terraform/aws/aws.tf]: you have access_port/access_scheme next to each other above, either follow here, or just realphabetize the lists

File: ansible/roles/k8s-proxy/defaults/main.yaml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+---
+proxy_type: node
+proxy_settings:
+  alsologtostderr: false
+  kubeconfig: "{{ kubernetes_cert_dir }}/kube_proxy/kubeconfig"
+  log_dir: "{{ kubernetes_log_dir }}/proxy"
+  logtostderr: false
+  master: "{{access_scheme}}://{{ master_private_ip }}:{{access_port}}"
```
Comment (spiffxp) [inline ansible/roles/k8s-proxy/defaults/main.yaml]: what drove the decision to rename `master_scheme` to `access_scheme`? ditto for `_port`

File: terraform/local/local.tf [hunk 8]
```diff
@@ -154,17 +157,17 @@ resource "execute_command" "command" {
...
   }
 
   provisioner "local-exec" {
-    command = "ANSIBLE_HOST_KEY_CHECKING=False ansible-playbook -i '${path.module}/rendered/ansible.inventory' --connection=local ${path.module}/../../ansible/generate_local_kubeconfig.yaml"
+    command = "ANSIBLE_HOST_KEY_CHECKING=False ansible-playbook -i '${path.module}/rendered/hosts' --connection=local ${path.module}/../../ansible/generate_local_kubeconfig.yaml"
```
Comment (spiffxp) [inline terraform/local/local.tf]: should this be using `${var. ansible_playbook_command}`?

File: ansible/roles/k8s-proxy/defaults/main.yaml (hunk 7 above)
Comment (maratoid) [inline ansible/roles/k8s-proxy/defaults/main.yaml]: seemed like a better name for something that pertains to master and api server

File: bin/kraken-connect.sh [hunk 9]
```diff
@@ -15,7 +15,9 @@ source "${my_dir}/utils.sh"
 src_cluster_dir="/kraken_data/${KRAKEN_CLUSTER_NAME}"
 containerfiles=(
   kraken_data:${src_cluster_dir}/ssh_config
-  kraken_data:${src_cluster_dir}/ansible.inventory
+  kraken_data:${src_cluster_dir}/hosts
+  kraken_data:${src_cluster_dir}/group_vars/cluster
+  kraken_data:${src_cluster_dir}/hosts
```
Comment (spiffxp) [inline bin/kraken-connect.sh]: dupe `/hosts` line

File: terraform/local/local.tf (hunk 8 above)
Comment (maratoid) [inline terraform/local/local.tf]: no. ansible_playbook_command is the kickoff for ansible-in-docker. this is something else. we could make a separate configurable variable for this too.

File: bin/kraken-connect.sh [hunk 10]
```diff
@@ -23,14 +25,19 @@ containerfiles=(
   ${KRAKEN_CONTAINER_NAME}:/opt/kraken/terraform/${KRAKEN_CLUSTER_TYPE}/${KRAKEN_CLUSTER_NAME}/terraform.tfvars
 )
 
+container_var_files=(
+  kraken_data:${src_cluster_dir}/group_vars/cluster
```
Comment (spiffxp) [inline bin/kraken-connect.sh]: this is mentioned in containerfiles as well, should it be?

File: terraform/aws/aws.tf [hunk 11]
```diff
@@ -654,32 +658,74 @@ resource "aws_route53_record" "proxy_record" {
 }
 
 resource "template_file" "ansible_inventory" {
```
Comment (spiffxp) [inline terraform/aws/aws.tf]: this should probably renamed to reflect that it's not a full inventory, just hosts?

Comment (maratoid) [inline terraform/aws/aws.tf]: it is still an 'Ansible inventory'. that's what it's called.

**Comment by spiffxp** [comment]
>breaks up the kubernetes ansible role into several smaller roles with dependencies through meta/main.yaml

this needs some kind of explanation in a README.md or something, for how/why they're broken up... doesn't have to be in this PR, because I expect things are still in flux, but this was a lot to digest

**Comment by maratoid** [comment]
retest this please

**Comment by maratoid** [comment]
retest this please

**Pull request merged**